### PR TITLE
Enable ko support to build & install

### PR DESCRIPTION
**Pull request by houshengbo**
# Changes
* This PR adds the ko support, so that operator can be installed with the command `ko apply -f config/`.
* We should set GO111MODULE=on, when in need. This should not work as the default option.
* A role for `rbac.authorization.k8s.io` is added for tekton-operator.

Fixes: #11 #40

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior

```

### CODE REVIEW

**Comment by houshengbo** [comment]
I will create the the new integration tests to verify if the CR can be installed, after operator is installed. Remove the testmain to make sure the integration test job is happy.

**Comment by sthaha** [comment]
@houshengbo @vdemeester @nikhil-thomas , Could we hold this PR until the switch away from `operator-sdk` can be discussed in a WG meeting and we get inputs from how knative operator is shipped into operatorhub.io without relying on operator-sdk and related tools?

**Comment by vdemeester** [comment]
/hold
@sthaha yes. I'm willing to prioritize the features (addons support, …), testing coverage and publishing a first version of the operator before change the tooling. 

Operator developers are either used to `operator-sdk` or `controller-runtime` more than `knative/pkg`. We might need to discuss that in a call (maybe we can setup a operator sync, the same way `tektoncd/triggers` does).

**Comment by houshengbo** [comment]
@vdemeester We can cancel the hold as well, since ko support does not influence the operator-sdk. It uses a different operator.yaml under config/.

File: config/operator.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  template:
+    metadata:
+      labels:
+        name: tekton-operator
+    spec:
+      serviceAccountName: tekton-operator
+      containers:
+        - name: tekton-operator
+          image: github.com/tektoncd/operator/cmd/manager
+          imagePullPolicy: IfNotPresent
```
Comment (vdemeester) [inline config/operator.yaml]: Not sure about this one :angel: (we don't set this on pipeline, …)

Comment (houshengbo) [reply config/operator.yaml]: removed it.

File: test/testgroups/cluster_crd.go [hunk 2]
```diff
@@ -3,50 +3,15 @@ package testgroups
...
 	defer ctx.Cleanup()
 
-	err := deployOperator(t, ctx)
-	helpers.AssertNoError(t, err)
-
 	t.Run("auto-installs-pipelines", testsuites.ValidateAutoInstall)
 	t.Run("delete-pipelines", testsuites.ValidateDeletion)
 }
-
-func deployOperator(t *testing.T, ctx *test.TestCtx) error {
```
Comment (vdemeester) [inline test/testgroups/cluster_crd.go]: Why is this removed ?

Comment (houshengbo) [reply test/testgroups/cluster_crd.go]: function tekton_setup() in the script will launch the operator, so we do not need to deploy the operator here.

File: deploy/operator.yaml [hunk 3]
```diff
@@ -15,11 +15,8 @@ spec:
       serviceAccountName: tekton-operator
```
Comment (vincent-pli) [inline deploy/operator.yaml]: Should restructure the `/deploy` or `/config`,  add number as prefix of files, like 500-operator.yaml, it will keep the order when run `ko apply`,  avoid exception caused by the wrong order.

File: README.md [hunk 4]
```diff
@@ -9,7 +9,7 @@ The following steps will install [Tekton Pipeline](https://github.com/tektoncd/p
     
 2. Apply Operator CRD
 
-    `kubectl apply -f deploy/crds/*_crd.yaml`
+    `kubectl apply -f deploy/crds/`
```
Comment (vincent-pli) [inline README.md]: If we turn to `ko`, the `crd` should also handle by `ko apply`, so the `crd` definition file should put in `/deploy/`

Comment (houshengbo) [reply README.md]: We do not need to use ko on top of the crds, because there is no image building or pulling involved so far.

File: deploy/operator.yaml (hunk 3 above)
Comment (houshengbo) [reply deploy/operator.yaml]: So far, the CRD installation still goes before the operator installation as prerequisite. We may need to do the ordering later if we decide to move the CRD into the operator installation as well.

File: README.md (hunk 4 above)
Comment (nikhil-thomas) [reply README.md]: @vincent-pli @houshengbo
I agree with vincent-pli
as we are providing `ko` support, it will be better if we can move all manifests to a single location and number them.

my only concern was the `operator-sdk olm-catalog gen-csv` command which generates a ClusterServiceVersion (CSV) file to be published on OperatorHub. By default, the command assumes the operator.yaml to be in deploy/ and crds in deploy/crd/.

However, the command can accept a config file where we can specify the paths,
eg: https://github.com/operator-framework/operator-sdk/blob/master/internal/pkg/scaffold/olm-catalog/testdata/deploy/olm-catalog/csv-config.yaml

Comment (nikhil-thomas) [reply README.md]: @houshengbo could you make this update

in fact, if we decide to move to knative-pkg injection-controller, this part can stay the same.

Comment (nikhil-thomas) [reply README.md]: >  there is no image building or pulling involved so far.

true, but if we keep them separate we will have to do a `kubectl apply -f <crd path>` and `ko apply config`
i would prefer haveing a single directory with numbered files

Comment (houshengbo) [reply README.md]: @nikhil-thomas I can use the file link instead of the physical file. We can still keep crd under deploy/crds, but file-link it under config. openrator-sdk works with deploy dir and ko works with config dir.

File: test/e2e-tests.sh [hunk 5]
```diff
@@ -14,15 +14,35 @@
...
+  header "Installing Tekton Operator"
+  kubectl create namespace $TEST_NAMESPACE
+  kubectl apply -n $TEST_NAMESPACE -f deploy/crds/
+  ko apply -n $TEST_NAMESPACE -f config/
+  wait_until_pods_running $TEST_NAMESPACE
+}
+
+function tekton_teardown() {
+  header "Removing Tekton Operator"
+  kubectl delete --ignore-not-found Config cluster
```
Comment (nikhil-thomas) [inline test/e2e-tests.sh]: use fully qualified name
`config.opeator.tekton.dev`

File: test/e2e-tests.sh [hunk 6]
```diff
@@ -14,15 +14,35 @@
...
-export GO111MODULE=on
 source $(dirname $0)/../vendor/github.com/tektoncd/plumbing/scripts/e2e-tests.sh
 
+# Namespace used for tests
+readonly TEST_NAMESPACE="operator-tests"
+
+function tekton_setup() {
+  header "Installing Tekton Operator"
+  kubectl create namespace $TEST_NAMESPACE
+  kubectl apply -n $TEST_NAMESPACE -f deploy/crds/
```
Comment (nikhil-thomas) [inline test/e2e-tests.sh]: why do we need `-n $TEST_NAMESPACE` while applying CRDs as CRDs are clusterwide

File: test/e2e-tests.sh [hunk 7]
```diff
@@ -14,15 +14,35 @@
...
+  ko apply -n $TEST_NAMESPACE -f config/
+  wait_until_pods_running $TEST_NAMESPACE
+}
+
+function tekton_teardown() {
+  header "Removing Tekton Operator"
+  kubectl delete --ignore-not-found Config cluster
+  ko apply -n $TEST_NAMESPACE -f config/
+  kubectl delete -n $TEST_NAMESPACE -f deploy/crds/
```
Comment (nikhil-thomas) [inline test/e2e-tests.sh]: why do we need `-n $TEST_NAMESPACE` while applying CRDs as CRDs are clusterwide

File: test/e2e-tests.sh [hunk 8]
```diff
@@ -14,15 +14,35 @@
...
+  kubectl create namespace $TEST_NAMESPACE
+  kubectl apply -n $TEST_NAMESPACE -f deploy/crds/
+  ko apply -n $TEST_NAMESPACE -f config/
+  wait_until_pods_running $TEST_NAMESPACE
+}
+
+function tekton_teardown() {
+  header "Removing Tekton Operator"
+  kubectl delete --ignore-not-found Config cluster
+  ko apply -n $TEST_NAMESPACE -f config/
```
Comment (nikhil-thomas) [inline test/e2e-tests.sh]: is this necessary in tear_down

File: test/e2e-tests.sh (hunk 6 above)
Comment (nikhil-thomas) [reply test/e2e-tests.sh]: or should it be ` kubectl apply -n $TEST_NAMESPACE -f deploy/crs/`

**Comment by nikhil-thomas** [comment]
@houshengbo 
as you said you might be going to fix it in a different patch.
However, `./test/e2e-tests.sh`
results in
```
error: set $PROJECT_ID or use --gcp-project to select the GCP project where the tests are run
```

**Comment by houshengbo** [comment]
@nikhil-thomas per ./test/e2e-tests.sh, this is an existing old issue with this script all across knative and tekton projects. We have to set an env var PROJECT_ID or --gcp-project to run the script, no matter we use google cloud or not.      I will get to it to see, if we can resolve it in a wide range.

File: test/e2e-tests.sh (hunk 6 above)
Comment (houshengbo) [reply test/e2e-tests.sh]: Since I create the CRDs available under config/, I can use kubectl apply -n $TEST_NAMESPACE -f config/ to install everything, including the crds and the deployment. So I can drop this line.

File: test/e2e-tests.sh (hunk 7 above)
Comment (houshengbo) [reply test/e2e-tests.sh]: Again, I have the link files of crds under config/, the crds will be removed if I run kubectl delete -n $TEST_NAMESPACE -f config/. I will change the tear_down as well.

File: test/e2e-tests.sh (hunk 8 above)
Comment (houshengbo) [reply test/e2e-tests.sh]: This is my fault, it should be delete.

File: test/e2e-tests.sh (hunk 5 above)
Comment (houshengbo) [reply test/e2e-tests.sh]: done

File: test/e2e-tests.sh (hunk 7 above)
Comment (nikhil-thomas) [reply test/e2e-tests.sh]: > @nikhil-thomas per ./test/e2e-tests.sh, this is an existing old issue with this script all across knative and tekton projects. We have to set an env var PROJECT_ID or --gcp-project to run the script, no matter we use google cloud or not. I will get to it to see, if we can resolve it in a wide range.

it set the env var PROJECT_ID, but then i need `gcloud` binary. In addtion, I have to provide a valid PROJECT_ID not any random value. :(

is that how it works at present

**Comment by nikhil-thomas** [comment]
> @nikhil-thomas per ./test/e2e-tests.sh, this is an existing old issue with this script all across knative and tekton projects. We have to set an env var PROJECT_ID or --gcp-project to run the script, no matter we use google cloud or not. I will get to it to see, if we can resolve it in a wide range.

it set the env var PROJECT_ID, but then i need gcloud binary. In addtion, I have to provide a valid PROJECT_ID not any random value. :(

is that how it works at present

File: test/e2e-tests.sh (hunk 7 above)
Comment (nikhil-thomas) [reply test/e2e-tests.sh]: it  makes more sense now :+1:

**Comment by houshengbo** [comment]
@nikhil-thomas It seems so at this moment. To run the complete script, we need to set up a google cloud account, and associate with the billing account. 

However, you can verify the script running, by the integration tests: https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_operator/21/pull-tekton-operator-integration-tests/1184851383740796930/. All the google clusters with credentials have been set up there, so the script can run the operator installation and the tests against a kube cluster in google cloud.
